### ISSUE TITLE
haxe 3.4.0 compilation freezing

### ISSUE THREAD

**Issue by r3d9u11** [HaxeFoundation/haxe#5918]
Hi. I have some troubles with haxe 3.4.0: when I try to compile openfl project with several compies of one font-file in different folders then compilation process will be freezed.

with haxe 3.2.1 this appliation will be already compiled.

tested OS: Windows 7 x64, Ubuntu 16.04 x64, Linux Mint 18 x64, Linux Mint 18.1 x64

Also this situation was described here: https://github.com/r3d9u11/StablexUI-Designer/issues/1

Thanks.

----
UPD:
example: [test.zip](https://github.com/HaxeFoundation/haxe/files/681864/test.zip)
will be compiled with haxe 3.2.1, but not with haxe 3.4.0

**Comment by Simn** [comment]
I can never remember now to compile these things...

**Comment by jgranick** [comment]
I believe there was some kind of recursion in how the macros were typed since the Haxe 3.4 RC releases, which did not exist in Haxe 3.2.1. By making sure that we do not reference classes that have `@:build` macros directly (or indirectly) in other macros, we have been able to avoid this issue.

This has been worked around in the latest Lime development builds.

For the record, this is all that is required :smile:

```bash
haxelib install openfl
haxelib run openfl setup
...
cd path/to/project
openfl test neko
```

**Comment by jgranick** [comment]
I do not have a simple reproduction case, but for reference, the issue is caused by something similar to this:

```haxe
@:build(CustomMacro.build()) class LimeFont {}
class Font extends LimeFont {}
class Arial extends Font {}
class Tahoma extends Font {}
```

For some reason, have more than one class that extends `openfl.text.Font` (which in turn extends `lime.text.Font`) caused the compiler to hang forever. Only one of these instances and it work properly.

Our workaround was to remove the `@:build` macro on `lime.text.Font`

There is an `@:autoBuild` on `lime.text.Font` as well, but that's unrelated -- commenting that out would still incur a freeze.

**Comment by matrefeytontias** [comment]
Bump,

I've ran into the same issue, but I have no working fix. @jgranick removing the `@:build` macro in `lime.text.Font` only makes it so that the CFFI native functions don't have a body, thus not actually compiling (but hey at least it doesn't hang I guess).

**Comment by oprudkyi** [comment]
I've the same issue (though with other library), I've tried to trace it down. as for me it seems like 						
Context.resolveType sometimes breaks and starts endless recursive loop for simple cases like 
```public var stringVar:String;```

i.e. in macro I got  t = FVar(TPath({ name => String, pack => [], params => [] }),null) and then calling to 
```Context.resolveType(t,f.pos);```
or
t.ToType() 
and instead returning value it just freezes (or returns to process other lines via recursion) 

though I have this error only in single project, other similar projects compiles well

**Comment by oprudkyi** [comment]
Hi, any changes to have fix soon ? the bug is really annoying and hard to bypass with third-party libraries

**Comment by oprudkyi** [comment]
ok. it seems storing each class in own .hx file helps

**Comment by RealyUniqueName** [comment]
Should be fixed in 4.0.0-rc.3 because of #8284
At least I cannot reproduce it with the latest dev Haxe
